internal/common/retry/interceptor.go

```diff
@@ -0,0 +1,146 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package retry
+
+import (
+	"context"
+	"math"
+	"time"
+
+	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/backoffutils"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+)
+
+const (
+	// NoMaximumAttempts is a value that maximumAttempts can be set to if total number of attempts should be unlimited.
+	NoMaximumAttempts = 0
```

Comment:
This is pretty confusing. It's saying "If you specify maximum attempts equal to anything that is the same as this constant, then max attempts is ignored?" 

A better name might be `IgnoreMaximumAttemptsSigil` with the docstring like "When maximum attempts is set to this special value, then the number of attempts is unlimited"

Same for `NoInterval`

But, the other problem here is negative values are nonsense. So perhaps this should be a `>=` check.

Reply:
I also don't think these docstrings need to repeat the field name (same with a lot of the other ones in this file)

Reply:
As suggested above, let's rename it to `UnlimitedMaximumAttempts`. Updated docstrings too.
Go lint actually requires you to repeat the name, kind of silly, but it's a hard requirement, which a lot of people don't like.

Reply:
Also I was considering making default values -1 instead of 0, which would make more sense, as 0 technically is a valid value, but I didn't want to introduce regression into existing retrier, that also uses these constants.